File: src/chain-observers/classifiers/manabar-classifier.ts
import { EManabarType, TAccountName } from "@hiveio/wax";
import { IMaxManabarData } from "./account-classifier";
import { CollectorClassifierBase, TRegisterEvaluationContext } from "./collector-classifier-base";

export interface IManabarDataPercent extends IMaxManabarData {
  percent: number;
}

export type TManabars = Partial<Record<EManabarType, IManabarDataPercent>>;

export interface IManabarAccountData {
  manabarData: Record<string, TManabars>;
}

export interface IManabarCollectorOptions {
  account: TAccountName;
  manabarType: EManabarType;
}

export class ManabarClassifier extends CollectorClassifierBase<{}, IManabarAccountData, void, void, IManabarCollectorOptions> {
  public static forOptions(options: IManabarCollectorOptions): TRegisterEvaluationContext {
    return {
      class: this, // Intentionally using `this` to refer to the class prototype itself later - even though it is not a class **instance**
      options
    };
  }
}

File: python/workerbee/chain_observers/classifiers/manabar_classifier.py
from __future__ import annotations

from typing import Any

from ..classifier_results import ManabarClassifierData
from .collector_classifier_base import CollectorClassifierBase, TRegisterEvaluationContext


class ManabarClassifier(CollectorClassifierBase[ManabarClassifierData]):
    """Collect account manabar data.

    Mirrors the TypeScript ``IManabarAccountData`` shape::

        IManabarAccountData:
            manabar_data: dict[str, TManabars]

        TManabars = dict[ManabarType, IManabarDataPercent]
        IManabarDataPercent:
            current_mana: int
            last_update_time: datetime
            max: int
            percent: float
    """

    @classmethod
    def for_options(cls, options: dict[str, Any]) -> TRegisterEvaluationContext:
        return {"class": cls, "options": options}
